### PR TITLE
Imu improvements

### PR DESCRIPTION
**Pull request by sfe-SparkFro**
A handful of general improvements to the IMU class. Notable changes include:
* Fix IMU calibration loop to actually get an average for the offset, rather than just the last measurement (`num_vals` was never incremented)
* Optimize IMU calibration to only divide once after measurement loop (division takes a long time)
* Add ODR functions
* Change default ODR to 208Hz, and default update period to 5ms (close as possible to 208Hz)
* Change default FS ranges to max (not clipping is more important than fine resolution, but feel free to change if max is too much)
* Check WHO_AM_I register during initialization (still need to implement something intelligent if it fails)
* Reset IMU during initialization (clears any previous configuration in case the processor is reset without power cycling the IMU)

### CODE REVIEW

**Comment by sfe-SparkFro** [comment]
Hold on, don't accept yet, just found a bug

**Comment by sfe-SparkFro** [comment]
Fixed!

**Review by Kreeevin — CHANGES_REQUESTED** [review]

File: XRPLib/imu.py
Comment (sfe-SparkFro) [reply XRPLib/imu.py:345]: First, to answer your questions - I was really just copying the structure of the scale functions. I'm actually not really keen on doing it like this; it's slow, takes more memory than it needs to, and it's not really how it should be done. Using numbers would be an improvement, but I think there's better solutions (eg. an enum).

Ok, this is where I tangent. This might turn into a longer discussion about the best way to architect the library. If so, it might be worth moving this conversation to its own issue in the repo. Sorry in advance 😉

Quick disclaimer: I mostly write C/C++ code (usually Arduino). There's actually a lot of things about Python that drive me crazy, so I don't use Python much. That means I'm also ignorant about how Python works under the hood, and I don't know a lot of features that I probably should. With that said:

If I were doing this in C/C++, there's some things I'd do differently:
1. Use enums for function parameters
    * Create one enum with each valid option for each function, then use those enums as the function parameters. That guarantees a correct value is passed at compile time, rather than having to check at runtime (faster execution, less program memory)
2. Use `#define`s for register names/addresses
    * This puts all those values into program memory rather than storing them in global variables, so this saves on RAM. Also, the compiler would just ignore any unused values, which further saves program memory
3. Use bitfield structs for register values ([explanation](https://en.cppreference.com/w/cpp/language/bit_field) for anyone who needs it)
    * This basically does all the bit shifting and masking for you. It guarantees that you don't accidentally set any bits that you didn't mean to, and actually results in faster code on some architectures

Again, I'm ignorant about Python, so I don't know the best practices for this kind of stuff. But after some Googling and testing, I have some thoughts:
1. Looks like enums currently [aren't support in MicroPython](https://github.com/micropython/micropython-lib/issues/269)
    * I'm not sure what the best alternative would be. I think the tuple solution that's being used now is interesting, where the index is the value for each entry. But I don't think it's very scalable, it's difficult to verify the indices are correct, and it requires adding blank entries to skip values. I think something like a dictionary might make more sense if we can't use enums.
    * This unfortunately doesn't guarantee the provided value is valid (insert grumbling about dynamic type casting and no compile time checks in Python...). Regardless of what we expect the parameters to be, it would be wise to validate all parameters of all functions (at least, any functions that users would call)
2. Use [MicroPython's `const`](https://docs.micropython.org/en/latest/library/micropython.html#micropython.const) for register names/addresses
    * This should help with RAM usage. Don't think it'll help much with program memory (arguably worse because there's more text in the .py file, but that's another conversation...), but probably worth it
3. Use [MicroPython's `uctype` bit field structs](https://docs.micropython.org/en/latest/library/uctypes.html) for register values
    * I've been tinkering with this a bit, and it behaves like I expect. It's not documented as well as I'd like, but it's relatively straight forward I think.

If you need, I might be able to provide an example that includes the ideas I'm suggesting here. Let me know your thoughts!

File: XRPLib/imu.py
Comment (sfe-SparkFro) [reply XRPLib/imu.py:431]: I would actually refrain from implementing a gravitational offset in the library, at least by default. Gravity can be really useful for determining the pitch and roll of the robot, freefall detection, etc. Plus it's really easy for the user to implement this offset by themselves if they need it.

File: XRPLib/imu.py
Comment (sfe-SparkFro) [reply XRPLib/imu.py:67]: This probably ties into the other comment I made regarding the best way to architect the library. In general, what should the library do when an error occurs? I would argue that throwing exceptions would be best when the problem can't be easily solved. We could even define custom exceptions that spell out to the user what went wrong, so no need to print anything.

File: XRPLib/imu.py
Comment (Kreeevin) [reply XRPLib/imu.py:431]: The gravitational offset here is because we initially assume all values are 0 by default when taking the averages that we use for the offets. We then have to go back and correct one of the axis so that the gravity vector is actually visible in the readings, otherwise all we'll get after calibration is zeros

File: XRPLib/imu.py
Comment (Kreeevin) [reply XRPLib/imu.py:345]: i agree with a lot of what you said here; I've used pretty much this exact IMU in C++, and i did basically that (except the bitfields, which i've never messed with). 

In regards to the Micropython stuff, enums not being supported is pretty annoying. There's definitely still ways around that though, and I'm tempted to go and try to make that work anyways. I definitely agree with parameter validation for functions where the type of which isn't explicitly defined (or at least add much better documentation). 

I had originally had it set up using `const`, but that was causing issues with Sphinx and the auto-documentation generation. If RAM usage winds up being an issue I can go back and revisit trying to get that to work.

As I mentioned, i haven't used bitfields before so that'll be new, but it definitely seems like a worthwhile thing to look into in terms of cleaning up this code a little. 

Overall I just want to make sure that the API is as easy to use as possible. I really would like to at the very least shift over to an enum-like situation, even if it's not as efficient as it would be had micropython allowed it. I'm thinking maybe just a dictionary would be a little better?

Let me know if you want to give it a shot, otherwise I'm fine merging your stuff in now and having me go back to clean up everything later

File: XRPLib/imu.py
Comment (Kreeevin) [reply XRPLib/imu.py:67]: So far I've been running by the idea that errors should be caught somewhere in the library if they can happen. So far I've been running with QoL as my number 1 priority in terms of making sure that something like the ultrasonic timing out doesn't lead to everything else breaking. There really aren't many situations where an exception can be thrown from the library side, but so far I've been leaning in favor of catching errors and then returning an error code esque value when something goes wrong (like returning 65535 for distance if it times out) Definitely not sure that's the best solution, so lmk what you think

File: XRPLib/imu.py
Comment (sfe-SparkFro) [reply XRPLib/imu.py:345]: Ok, so here's an example of how I think we could improve the structure of the IMU code, implementing the 3 things I mentioned (`const`, dictionaries, and bit field structs). This should be pretty easy to scale up as needed. This code here is just a bunch of definitions, so it'd maybe make sense to put it in its own `imu_defs.py` file or something:

```
from uctypes import BFUINT8, BF_POS, BF_LEN, struct, addressof
from micropython import const

"""
	Register addresses
"""
LSM_REG_CTRL1_XL = const(0x10)
LSM_REG_CTRL2_G  = const(0x11)
LSM_REG_CTRL3_C  = const(0x12)
LSM_REG_CTRL4_C  = const(0x13)
LSM_REG_CTRL5_C  = const(0x14)
LSM_REG_CTRL6_C  = const(0x15)
LSM_REG_CTRL7_G  = const(0x16)
LSM_REG_CTRL8_XL = const(0x17)
LSM_REG_CTRL9_XL = const(0x18)
LSM_REG_CTRL10_C = const(0x19)

"""
	Dictionaries for possible register settings
"""
LSM_ODR = {
	"0Hz"    : 0x0,
	"12.5Hz" : 0x1,
	"26Hz"   : 0x2,
	"52Hz"   : 0x3,
	"104Hz"  : 0x4,
	"208Hz"  : 0x5,
	"416Hz"  : 0x6,
	"833Hz"  : 0x7,
	"1660Hz" : 0x8,
	"3330Hz" : 0x9,
	"6660Hz" : 0xA,
}
LSM_ACCEL_FS = {
	"2g"  : 0x0,
	"16g" : 0x1,
	"4g"  : 0x2,
	"8g"  : 0x3,
}
LSM_GYRO_FS = {
	"125dps"  : 0x1,
	"250dps"  : 0x0,
	"500dps"  : 0x2,
	"1000dps" : 0x4,
	"2000dps" : 0x6,
}

"""
	Bit field struct definitions of registers
"""
LSM_REG_CTRL1_XL_LAYOUT = {
    "LPF2_XL_EN" : BFUINT8 | 1 << BF_POS | 1 << BF_LEN,
    "FS_XL"      : BFUINT8 | 2 << BF_POS | 2 << BF_LEN,
    "ODR_XL"     : BFUINT8 | 4 << BF_POS | 4 << BF_LEN,
}
LSM_REG_CTRL2_G_LAYOUT = {
    "FS_G"  : BFUINT8 | 1 << BF_POS | 3 << BF_LEN,
    "ODR_G" : BFUINT8 | 4 << BF_POS | 4 << BF_LEN,
}

"""
	Copies of register values
"""
reg_ctrl1_xl_byte   = bytearray(1)
reg_ctrl1_xl_struct = struct(addressof(reg_ctrl1_xl_byte), LSM_REG_CTRL1_XL_LAYOUT)
reg_ctrl2_g_byte   = bytearray(1)
reg_ctrl2_g_struct = struct(addressof(reg_ctrl2_g_byte), LSM_REG_CTRL2_G_LAYOUT)
```

Obviously this doesn't include everything, just getting the idea across. If I were doing this in C/C++, I'd add everything for every register. However we need to keep in mind file size, since all of this text gets stored on the Pico W, which only has 2MB of flash. So it would probably be better in this case to only add the things we actually use.

Then in the IMU class, the member functions would look more like this:

```
def acc_rate(self, dat=None):
	"""
	Set the accelerometer rate. The rate can be '0Hz', '12.5Hz', '26Hz', '52Hz', '104Hz', '208Hz', '416Hz', '833Hz', '1660Hz', '3330Hz', '6660Hz'.
	Pass in no parameters to retrieve the current value
	"""
	# Check if dat is valid input
	if (dat is None) or (type(dat) is not str) or (dat not in LSM_ODR):
		# Not valid input, return ODR value
		reg_ctrl1_xl_byte = self._getreg(LSM_REG_CTRL1_XL)
		return reg_ctrl1_xl_struct.ODR_XL
	else:
		# Valid input, update ODR value
		reg_ctrl1_xl_byte = self._getreg(LSM_REG_CTRL1_XL)
		reg_ctrl1_xl_struct.ODR_XL = LSM_ODR[dat]
		return self._setreg(LSM_REG_CTRL1_XL, reg_ctrl1_xl_byte)
```

Note - I've not actually tested any of this, don't have my robot with me right now.

There's definitely pros and cons to this approach. The benefit is that this is pretty straight forward to scale up, there's no magic numbers, and it's quite readable IMO. The downside is that this takes up more memory, both flash and RAM (don't think dictionaries work with `const`). There's probably other considerations I'm not thinking of right now.

Let me know what you think!

File: XRPLib/imu.py
Comment (sfe-SparkFro) [reply XRPLib/imu.py:431]: Oh I see, I misunderstood what the code was doing, my bad!

Hmm, thinking about it, should we really include the accelerometer in the offset calibration? This assumes the board is perfectly flat with respect to gravity, but that definitely won't be the case for all users. How far off are the measurements with no offset? Is an offset actually needed? Can we perhaps separate the gryo and accelerometer calibration, then leave it up to the user to run the accelerometer calibration? Could create a separate `imu.calibrate_accel()` and allow the user to pass in the orientation of their board.

File: XRPLib/imu.py
Comment (sfe-SparkFro) [reply XRPLib/imu.py:67]: Good point, totally agree with QoL being priority. I think we should distinguish between critical and non-critical failures. I would argue that something like the ultrasonic sensor timing out would be non-critical, and shouldn't bring everything to a halt, whereas something like the IMU not being detected during initialization is a critical failure, and should stop everything to alert the user. We may have different opinions on what constitutes critical and non-critical failures, but I do think there are cases when everything needs to stop so the user is aware of the problem.

As for indicating non-critical errors, I'm hesitant about returning special numbers. Since we're targeting people who have little technical experience, I don't think we can't assume they know big powers of 2, and that big powers of 2 are sometimes used to indicate special information. Negative values are arguably a bit better, but that doesn't work for all cases (sometimes -1 is a valid value, for example), and still doesn't clearly convey the information. Maybe we can return something else entirely, like `None` or a string with the error in plain text? Or have some other member function like `get_error_code()` that returns the most recent non-critical failure that occurred? Idk, not sure what's best either.

Comment (Kreeevin) [reply XRPLib/imu.py:67]: If we're going to return a string as an error code, we might as well raise an exception. I'd be okay changing to -1 as a timeout code for ultrasonic, but part of the reason I had it at 65535 was so that if anybody was trying to drive until they were within a given range, a timeout wouldn't make them stop. I agree with critical errors raising exceptions, but the imu is technically a non-essential function since the drivetrain class doesn't actually need it to operate. That's why I'm caught between just raising an error or just letting it be None and giving a warning

Comment (Kreeevin) [reply XRPLib/imu.py:67]: maybe have the IMU raise an exception and the drivetrain get_default method catches it if we decide to go down that route?

File: XRPLib/imu.py
Comment (Kreeevin) [reply XRPLib/imu.py:431]: There's definitely a non-negligible offset from the accelerometer readings. It doesn't need to be as carefully calibrated as the gyro bc we're not really integrating it, so drift doesn't really matter, but we should definitely be calibrating it at least a little before letting the students use it.

Separating the gyro and accelerometer calibration is an interesting idea, but passing in the orientation (or using the gyro) is still based in them having calibrated the gyro on a flat surface. I honestly think it's probably fine to say that most people will be calibrating their bots on a flat (or almost flat) surface, and that the slight deviation from that will have a relatively small impact.

The current implementation allows for people to calibrate their board using either the X, Y, or Z axis as the one starting with gravity positive, but the more I think about it the less use cases I can think for this option. It may just be easier (and less confusing to someone looking at the documentation) if we were to just remove that option

File: XRPLib/imu.py
Comment (Kreeevin) [reply XRPLib/imu.py:345]: How much of the flash memory does the current library actually take up? I can't imagine adding the dictionaries would have a substantial impact there. I think in terms of clean implementation, I like the dictionaries a lot better than the current solution, since at least then the code becomes readable for any student looking to go above and beyond and try to understand how this stuff works (expecting this moreso from the WPI students using these in a class, not for the high schoolers). I also agree with moving these into a new file to keep `imu.py` clean.

If you want to implement that stuff as part of this PR, that would be awesome

**Comment by Kreeevin** [comment]
Another comment for a pretty big improvement to this code; now that we're providing a way to change the data rates, we probably don't just want to be assuming that the imu update time is 5ms (which is an approximation all on it's own). What you can do is change the update timer to use frequency instead of period (set after calibration) and then pass in the currently set ODR for the gyro, which would eliminate any drift that would have resulted from using the wrong frequency. That would probably be a good change to make before this gets merged in

File: XRPLib/imu.py
Comment (sfe-SparkFro) [reply XRPLib/imu.py:345]: I believe the library currently uses about 200kB, including phew. I just downloaded the latest version of MicroPython for the Pico W, and it's 1368kB, so that presumably means we're using about 75% of the 2MB flash. We'll need to make sure there's extra room for user code, plus any updates to MicroPython and this library. Adding every definition for every register can definitely take a lot of space; [this file from ST's API](https://github.com/stm32duino/LSM6DSO/blob/main/src/lsm6dso_reg.h) is over 100kB of text, for example. Obviously we don't need to go all out like that, but I just want to make sure we're not being egregious with the amount of storage we take up.

I agree, I definitely prefer the code to be clean at the cost of some extra storage space. I'll start implementing this new code structure, and we'll see how much space it actually uses up. If needed, I think we can just delete any portions that we're not using to save space.

File: XRPLib/imu.py
Comment (sfe-SparkFro) [reply XRPLib/imu.py:67]: All good points! I guess one other consideration we should make is how frequently we expect these problems to occur. For example, I expect the IMU to fail very rarely (it's soldered directly to the board and tested in production), whereas something like the ultrasonic sensor is much more likely to go wrong (users may forget to plug the sensor in, or wire it up incorrectly, or there's an intermittent connection, or...).

As for how to indicate problems, I don't know if there's a "best" solution here, each seems to have a downside. But after thinking about it some more, I'm personally leaning towards throwing exceptions in most cases. Exceptions explain the problem to the user in plain text, so there's no ambiguity about what the problem is. If it's something that's easily fixed (eg. plug in the ultrasonic cable), then the user can quickly fix it and keep going. If it's something that's not easily fixed (eg. the IMU fails to communicate), the user should really be informed about it. Plus, it's pretty easy for the user to catch any exceptions that come up, and I think it'd be good for students to learn about exceptions at some point.

Truthfully, I don't really have any strong opinions on the best way to indicate problems. Each has their benefits and drawbacks, so I don't think any of them are bad choices. I don't think I have anything more to say on the subject, so I'll leave it up to you!

File: XRPLib/imu.py
Comment (sfe-SparkFro) [reply XRPLib/imu.py:431]: Fair points! I had a couple other thoughts come to mine:

Do we need to calibrate the IMU every single time, or can we do it once and store the offsets in a file or something? That'd help reduce the down time before user code executes, and make the calibration less prone to the user moving the robot during the calibration.

On a related note, what if the user doesn't know about the calibration routine, and they move the robot during it? I suppose we could detect it by evaluating the accel/gyro values, but then what? I'm wondering if it would be better to have the user call the calibrate function themselves, or something that bring the users attention to the fact that they shouldn't move the robot during that time.

I also think it may be worth revisiting this topic if/when an AHRS algorithm is implemented. In that case, the accelerometer would be used to determine the pitch and roll angles, so might affect how we decide to calibrate it (eg. if the user starts the robot on a surface tilted at 10 degrees, should the IMU say that's 10 degrees or zero degrees?)

**Comment by sfe-SparkFro** [comment]
Good point, I can change the timer to use `frequency` instead of `period` (didn't even realize that was an option!)

**Comment by fgrossman** [comment]
While talking about improvements to main.py.  We could go to blink codes on
the LED for certain exceptions. That way if something happens when they are
not attached to the computer, like a wire coming off the rangefinder, there
is a chance they could fix it.

     Frank

File: XRPLib/imu.py
Comment (sfe-SparkFro) [reply XRPLib/imu.py]: Just pushed this new structure. Haven't tested it yet, so hopefully it's not completely broken 😉 I decided to only include the registers we're actually using, since adding everything increases both file size and RAM usage (more grumbling about Python...). But it should be easy to expand as needed.

Let me know what you think!

**Comment by sfe-SparkFro** [comment]
I like the idea of blink codes in main.py!

I'm cleaning up some of the IMU code right now. Do we really need `power()`? There's some edge cases that could occur with the rate functions (eg. if a user calls `acc_rate(rate_1)`, then `power(False)`, then `acc_rate(rate_2)`, then `power(True)`, the rate will go back to `rate_1`). Obviously we could handle the edge cases, but I'm not really sure what the use case is for including `power()` with the XRP. Everything else takes up way more power, and if the user really wants to turn the IMU off, they can just send '0Hz' to the rate functions.

**Comment by sfe-SparkFro** [comment]
Ok, finally got around to testing this. Had to fix a couple mistakes, but should be all working now!

File: XRPLib/imu.py
Comment (Kreeevin) [reply XRPLib/imu.py:61]: I think that all makes sense, yeah. Going to resolve this comment for now, I'll message in Slack if I have any other questions on how to best improve anything

File: XRPLib/imu.py
Comment (Kreeevin) [reply XRPLib/imu.py]: I definitely like the organization of the imu_defs file a lot more than just having constants at the top of the imu class, thanks!

and yea, I don't anticipate really needing access to every single register option, what's there now should suffice.

File: XRPLib/imu.py
Comment (Kreeevin) [reply XRPLib/imu.py:517]: Storing calibration data is an interesting idea, definitely worth exploring it. 

Atm we're calibrating the gyro immediately for ease of use, since the default functionality is for the drivetrain to use it. We could change that, but I think it's useful for the drivetrain to just work when using it. What if we were to move the calibrate call from `get_default_imu` to `get_default_drivetrain`, so if a user wanted to use the IMU on its own they can control when exactly it calibrates? although they can just call calibrate on their own anyways. 

I'm not sure users knowledge of the calibration routine is a huge issue, since the bots are being coupled with a class and we can just makes sure the teachers pass that knowledge along. Would it be worthwhile to add a print statement or something that just says "calibrating robot, please don't move the robot during this time" and "Finished calibration, program running"?

As for that last note, I'm not sure which behavior is preferable. I've asked a few people what their personal thoughts are and overall I'm thinking that when you calibrate, all of your angles should be relative to that point. I'm going to be trying to implement the AHRS algorithm fairly soon, so we can revisit that discussion then.

File: XRPLib/imu.py
Comment (Kreeevin) [inline XRPLib/imu.py:115]: Very minor thing, just for organization purposes, is_connected and reset should be moved down to below the comment labeling that section as public facing methods

File: XRPLib/imu.py
Comment (Kreeevin) [inline XRPLib/imu.py:125]: Reset should probably stop the update timer and clear all of the offset values

**Review by Kreeevin — CHANGES_REQUESTED** [review]
Just a few minor nitpicks, everything looks pretty good otherwise

File: XRPLib/imu.py
Comment (sfe-SparkFro) [reply XRPLib/imu.py:115]: Woah! I'm blind, didn't realize the functions were divided up 😅 I'll move those into the public section.

While we're on this topic, any reason `_get_gyro_x_rate()` (y and z too), `_get_gyro_rates()`, and `_get_acc_gyro_rates()` can't be public? `get_acc_x()` (y and z too) and `get_acc_rates()` are public. I understand that users will usually get gyro data from `get_yaw()` (and others), but it feels weird to me for the accelerometer functions to be public, and the gyro functions to be private, especially since they're all very similar to each other. Eh, probably doesn't matter, but figured I'd suggest it.

Also just noticed `get_acc_rates()` probably shouldn't have `rate` in the name, since it's not a rate like the gyro measurements are.

File: XRPLib/imu.py
Comment (sfe-SparkFro) [reply XRPLib/imu.py:125]: Good call! Will add that

File: XRPLib/imu.py
Comment (sfe-SparkFro) [reply XRPLib/imu.py:527]: My main concern is that many (most?) people won't even know there's a calibration that occurs, and that the robot needs to be completely motionless during it. I could totally foresee a lot of people moving the robot around during that time for whatever reason (eg. carrying the robot while turning it on, or positioning it on a line just after clicking the run button in the IDE, or lightly bumping the robot, or...), then having their robot not drive straight or not turn to the correct angles. That will inevitably cause confusion, and will probably make some people think the robot is just bad, or theirs is broken. I can't count how many times I've seen hidden behavior like this trip up students.

I wouldn't require teachers relay that information (they've got enough to do already), plus there's a lot more use cases for the XRP than just classes. IMO it should be trivial for a user to discover that there is a calibration routine, and that the robot needs to be motionless during it. At a minimum, I think adding some prints statements before and after would be good. I think it'd also help to have it documented in multiple places (eg. the API documentation and the online curriculum). Or even better would be storing the calibration values in a file (could add a button in the IDE for this) so users don't have to worry about it all the time (assuming the IMU doesn't need to be calibrated frequently, need to test that).

Whether the angles should be relative or absolute with respect to gravity, this may actually depend on the AHRS algorithm. Most require the accelerometer ([list of algorithms here](https://ahrs.readthedocs.io/en/latest/#new-in-version-0-3)), and I would imagine those would bring the roll and pitch to zero if the accelerometer measures (x,y,z)=(0,0,1). I suppose you could add offsets for the roll and pitch if that's really desired. Idk if relative or absolute angles are best, but I'd personally lean towards absolute.

Also with regards to the accelerometer calibration, one other consideration is that the X and Y axis bounds should really be as close to +/- 1g as possible when the board is tilted sideways. After playing with my board, if the calibration is done on a non-level surface, the bounds can end up being wrong (eg. +1.1 and -0.9). So probably best to mention that the calibration should be done on a level surface, in addition to being motionless.

**Comment by fgrossman** [comment]
We could also add to the curriculum an IMU chapter, so they learn about IMU
calibration. And understand it better. Seems like something we would want
to teach.

File: XRPLib/imu.py
Comment (Kreeevin) [reply XRPLib/imu.py:527]: Yea currently calibrating on a non-flat surface is gonna cause some issues atm, since we do the `vertical_axis -= 1000`, which assumes that gravity is aligned perfectly with one of the vectors. I'm still not happy with this solution, so i'm tempted to just say that it has to be flat, and to remove the option to change the vertical axis. I think we can still revisit this later, when the AHRS code is ready to be merged.

Regardless, I'm okay going through and testing how well saving calibration data to a file works myself, so I think we're all good to merge your changes in now

File: XRPLib/imu.py
Comment (Kreeevin) [reply XRPLib/imu.py]: I agree with you here, I think they should be public. Ansel had made them private somewhere down the line but I think it's a better learning tool if we expose the fact that the gyro returns a rate instead of just the raw value; I just haven't gone through and made them public yet.

File: XRPLib/imu.py
Comment (sfe-SparkFro) [reply XRPLib/imu.py:527]: Sounds good! Will leave it as is for now so we can merge this.

File: XRPLib/imu.py
Comment (sfe-SparkFro) [reply XRPLib/imu.py:145]: Done!

**Review by Kreeevin — APPROVED** [review]
All functionality confirmed
Review complete

**Pull request merged**
